// @cloudflare/playwright/lib/cjs/playwright-core/src/server/dispatchers/artifactDispatcher.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const fs = require('../../../../bundles/fs.js');
const dispatcher = require('./dispatcher.js');
const streamDispatcher = require('./streamDispatcher.js');
const fileUtils = require('../utils/fileUtils.js');

class ArtifactDispatcher extends dispatcher.Dispatcher {
  constructor(scope, artifact) {
    super(scope, artifact, "Artifact", {
      absolutePath: artifact.localPath()
    });
    this._type_Artifact = true;
  }
  static from(parentScope, artifact) {
    return ArtifactDispatcher.fromNullable(parentScope, artifact);
  }
  static fromNullable(parentScope, artifact) {
    if (!artifact)
      return void 0;
    const result = dispatcher.existingDispatcher(artifact);
    return result || new ArtifactDispatcher(parentScope, artifact);
  }
  async pathAfterFinished() {
    const path = await this._object.localPathAfterFinished();
    return { value: path };
  }
  async saveAs(params) {
    return await new Promise((resolve, reject) => {
      this._object.saveAs(async (localPath, error) => {
        if (error) {
          reject(error);
          return;
        }
        try {
          await fileUtils.mkdirIfNeeded(params.path);
          await fs.default.promises.copyFile(localPath, params.path);
          resolve();
        } catch (e) {
          reject(e);
        }
      });
    });
  }
  async saveAsStream() {
    return await new Promise((resolve, reject) => {
      this._object.saveAs(async (localPath, error) => {
        if (error) {
          reject(error);
          return;
        }
        try {
          const readable = fs.default.createReadStream(localPath, { highWaterMark: 1024 * 1024 });
          const stream = new streamDispatcher.StreamDispatcher(this, readable);
          resolve({ stream });
          await new Promise((resolve2) => {
            readable.on("close", resolve2);
            readable.on("end", resolve2);
            readable.on("error", resolve2);
          });
        } catch (e) {
          reject(e);
        }
      });
    });
  }
  async stream() {
    const fileName = await this._object.localPathAfterFinished();
    const readable = fs.default.createReadStream(fileName, { highWaterMark: 1024 * 1024 });
    return { stream: new streamDispatcher.StreamDispatcher(this, readable) };
  }
  async failure() {
    const error = await this._object.failureError();
    return { error: error || void 0 };
  }
  async cancel() {
    await this._object.cancel();
  }
  async delete(_, metadata) {
    metadata.potentiallyClosesScope = true;
    await this._object.delete();
    this._dispose();
  }
}

exports.ArtifactDispatcher = ArtifactDispatcher;
