// @cloudflare/playwright/lib/cjs/playwright-core/src/server/dispatchers/jsHandleDispatcher.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const dispatcher = require('./dispatcher.js');
const elementHandlerDispatcher = require('./elementHandlerDispatcher.js');
const serializers = require('../../protocol/serializers.js');

class JSHandleDispatcher extends dispatcher.Dispatcher {
  constructor(scope, jsHandle) {
    super(scope, jsHandle, jsHandle.asElement() ? "ElementHandle" : "JSHandle", {
      preview: jsHandle.toString()
    });
    this._type_JSHandle = true;
    jsHandle._setPreviewCallback((preview) => this._dispatchEvent("previewUpdated", { preview }));
  }
  async evaluateExpression(params) {
    return { value: serializeResult(await this._object.evaluateExpression(params.expression, { isFunction: params.isFunction }, parseArgument(params.arg))) };
  }
  async evaluateExpressionHandle(params) {
    const jsHandle = await this._object.evaluateExpressionHandle(params.expression, { isFunction: params.isFunction }, parseArgument(params.arg));
    return { handle: elementHandlerDispatcher.ElementHandleDispatcher.fromJSHandle(this.parentScope(), jsHandle) };
  }
  async getProperty(params) {
    const jsHandle = await this._object.getProperty(params.name);
    return { handle: elementHandlerDispatcher.ElementHandleDispatcher.fromJSHandle(this.parentScope(), jsHandle) };
  }
  async getPropertyList() {
    const map = await this._object.getProperties();
    const properties = [];
    for (const [name, value] of map)
      properties.push({ name, value: elementHandlerDispatcher.ElementHandleDispatcher.fromJSHandle(this.parentScope(), value) });
    return { properties };
  }
  async jsonValue() {
    return { value: serializeResult(await this._object.jsonValue()) };
  }
  async dispose(_, metadata) {
    metadata.potentiallyClosesScope = true;
    this._object.dispose();
    this._dispose();
  }
}
function parseArgument(arg) {
  return serializers.parseSerializedValue(arg.value, arg.handles.map((a) => a._object));
}
function serializeResult(arg) {
  return serializers.serializeValue(arg, (value) => ({ fallThrough: value }));
}

exports.JSHandleDispatcher = JSHandleDispatcher;
exports.parseArgument = parseArgument;
exports.serializeResult = serializeResult;
