// @cloudflare/playwright/lib/cjs/playwright-core/src/server/dispatchers/tracingDispatcher.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const artifactDispatcher = require('./artifactDispatcher.js');
const dispatcher = require('./dispatcher.js');

class TracingDispatcher extends dispatcher.Dispatcher {
  constructor(scope, tracing) {
    super(scope, tracing, "Tracing", {});
    this._type_Tracing = true;
  }
  static from(scope, tracing) {
    const result = dispatcher.existingDispatcher(tracing);
    return result || new TracingDispatcher(scope, tracing);
  }
  async tracingStart(params) {
    await this._object.start(params);
  }
  async tracingStartChunk(params) {
    return await this._object.startChunk(params);
  }
  async tracingGroup(params, metadata) {
    const { name, location } = params;
    await this._object.group(name, location, metadata);
  }
  async tracingGroupEnd(params) {
    await this._object.groupEnd();
  }
  async tracingStopChunk(params) {
    const { artifact, entries } = await this._object.stopChunk(params);
    return { artifact: artifact ? artifactDispatcher.ArtifactDispatcher.from(this, artifact) : void 0, entries };
  }
  async tracingStop(params) {
    await this._object.stop();
  }
}

exports.TracingDispatcher = TracingDispatcher;
