// @cloudflare/playwright/lib/cjs/playwright-core/src/utils/isomorphic/builtins.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

function builtins(global) {
  global = global ?? globalThis;
  if (!global["__playwright_builtins__"]) {
    const builtins2 = {
      setTimeout: global.setTimeout?.bind(global),
      clearTimeout: global.clearTimeout?.bind(global),
      setInterval: global.setInterval?.bind(global),
      clearInterval: global.clearInterval?.bind(global),
      requestAnimationFrame: global.requestAnimationFrame?.bind(global),
      cancelAnimationFrame: global.cancelAnimationFrame?.bind(global),
      requestIdleCallback: global.requestIdleCallback?.bind(global),
      cancelIdleCallback: global.cancelIdleCallback?.bind(global),
      performance: global.performance,
      eval: global.eval?.bind(global),
      Intl: global.Intl,
      Date: global.Date,
      Map: global.Map,
      Set: global.Set
    };
    Object.defineProperty(global, "__playwright_builtins__", { value: builtins2, configurable: false, enumerable: false, writable: false });
  }
  return global["__playwright_builtins__"];
}
const instance = builtins();
const setTimeout = instance.setTimeout;
const clearTimeout = instance.clearTimeout;
instance.setInterval;
instance.clearInterval;
instance.requestAnimationFrame;
instance.cancelAnimationFrame;
instance.requestIdleCallback;
instance.cancelIdleCallback;
const performance = instance.performance;
instance.Intl;
instance.Date;
const Map = instance.Map;
const Set = instance.Set;

exports.Map = Map;
exports.Set = Set;
exports.builtins = builtins;
exports.clearTimeout = clearTimeout;
exports.performance = performance;
exports.setTimeout = setTimeout;
