// @cloudflare/playwright/lib/cjs/playwright-core/src/remote/playwrightConnection.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const socksProxy = require('../server/utils/socksProxy.js');
require('../server/registry/index.js');
const dispatcher = require('../server/dispatchers/dispatcher.js');
const playwrightDispatcher = require('../server/dispatchers/playwrightDispatcher.js');
const playwright = require('../server/playwright.js');
require('../../../bundles/fs.js');
require('node:path');
const assert = require('../utils/isomorphic/assert.js');
require('../utils/isomorphic/selectorParser.js');
require('../utils/isomorphic/builtins.js');
require('../utils/isomorphic/mimeType.js');
const time = require('../utils/isomorphic/time.js');
require('../utils/isomorphic/urlMatch.js');
require('../../../_virtual/pixelmatch.js');
require('../utilsBundle.js');
require('node:crypto');
const debug = require('../server/utils/debug.js');
const debugLogger = require('../server/utils/debugLogger.js');
require('../zipBundle.js');
require('../server/utils/hostPlatform.js');
require('node:http');
require('node:http2');
require('node:https');
require('node:url');
require('../server/utils/happyEyeballs.js');
require('../server/utils/nodePlatform.js');
require('node:child_process');
require('node:readline');
const profiler = require('../server/utils/profiler.js');
require('node:os');
require('../server/utils/zones.js');
const instrumentation = require('../server/instrumentation.js');
require('../protocol/serializers.js');
const android = require('../server/android/android.js');
const browser = require('../server/browser.js');
const debugControllerDispatcher = require('../server/dispatchers/debugControllerDispatcher.js');

class PlaywrightConnection {
  constructor(lock, clientType, ws, options, preLaunched, id, onClose) {
    this._cleanups = [];
    this._disconnected = false;
    this._ws = ws;
    this._preLaunched = preLaunched;
    this._options = options;
    options.launchOptions = filterLaunchOptions(options.launchOptions, options.allowFSPaths);
    if (clientType === "reuse-browser" || clientType === "pre-launched-browser-or-android")
      assert.assert(preLaunched.playwright);
    if (clientType === "pre-launched-browser-or-android")
      assert.assert(preLaunched.browser || preLaunched.androidDevice);
    this._onClose = onClose;
    this._id = id;
    this._profileName = `${(/* @__PURE__ */ new Date()).toISOString()}-${clientType}`;
    this._dispatcherConnection = new dispatcher.DispatcherConnection();
    this._dispatcherConnection.onmessage = async (message) => {
      await lock;
      if (ws.readyState !== ws.CLOSING) {
        const messageString = JSON.stringify(message);
        if (debugLogger.debugLogger.isEnabled("server:channel"))
          debugLogger.debugLogger.log("server:channel", `[${this._id}] ${time.monotonicTime() * 1e3} SEND ► ${messageString}`);
        if (debugLogger.debugLogger.isEnabled("server:metadata"))
          this.logServerMetadata(message, messageString, "SEND");
        ws.send(messageString);
      }
    };
    ws.on("message", async (message) => {
      await lock;
      const messageString = Buffer.from(message).toString();
      const jsonMessage = JSON.parse(messageString);
      if (debugLogger.debugLogger.isEnabled("server:channel"))
        debugLogger.debugLogger.log("server:channel", `[${this._id}] ${time.monotonicTime() * 1e3} ◀ RECV ${messageString}`);
      if (debugLogger.debugLogger.isEnabled("server:metadata"))
        this.logServerMetadata(jsonMessage, messageString, "RECV");
      this._dispatcherConnection.dispatch(jsonMessage);
    });
    ws.on("close", () => this._onDisconnect());
    ws.on("error", (error) => this._onDisconnect(error));
    if (clientType === "controller") {
      this._root = this._initDebugControllerMode();
      return;
    }
    this._root = new dispatcher.RootDispatcher(this._dispatcherConnection, async (scope, options2) => {
      await profiler.startProfiling();
      if (clientType === "reuse-browser")
        return await this._initReuseBrowsersMode(scope);
      if (clientType === "pre-launched-browser-or-android")
        return this._preLaunched.browser ? await this._initPreLaunchedBrowserMode(scope) : await this._initPreLaunchedAndroidMode(scope);
      if (clientType === "launch-browser")
        return await this._initLaunchBrowserMode(scope, options2);
      throw new Error("Unsupported client type: " + clientType);
    });
  }
  async _initLaunchBrowserMode(scope, options) {
    debugLogger.debugLogger.log("server", `[${this._id}] engaged launch mode for "${this._options.browserName}"`);
    const playwright$1 = playwright.createPlaywright({ sdkLanguage: options.sdkLanguage, isServer: true });
    const ownedSocksProxy = await this._createOwnedSocksProxy(playwright$1);
    let browserName = this._options.browserName;
    if ("bidi" === browserName) {
      if (this._options.launchOptions?.channel?.toLocaleLowerCase().includes("firefox"))
        browserName = "bidiFirefox";
      else
        browserName = "bidiChromium";
    }
    const browser$1 = await playwright$1[browserName].launch(instrumentation.serverSideCallMetadata(), this._options.launchOptions);
    this._cleanups.push(async () => {
      for (const browser2 of playwright$1.allBrowsers())
        await browser2.close({ reason: "Connection terminated" });
    });
    browser$1.on(browser.Browser.Events.Disconnected, () => {
      this.close({ code: 1001, reason: "Browser closed" });
    });
    return new playwrightDispatcher.PlaywrightDispatcher(scope, playwright$1, ownedSocksProxy, browser$1);
  }
  async _initPreLaunchedBrowserMode(scope) {
    debugLogger.debugLogger.log("server", `[${this._id}] engaged pre-launched (browser) mode`);
    const playwright = this._preLaunched.playwright;
    this._preLaunched.socksProxy?.setPattern(this._options.socksProxyPattern);
    const browser$1 = this._preLaunched.browser;
    browser$1.on(browser.Browser.Events.Disconnected, () => {
      this.close({ code: 1001, reason: "Browser closed" });
    });
    const playwrightDispatcher$1 = new playwrightDispatcher.PlaywrightDispatcher(scope, playwright, this._preLaunched.socksProxy, browser$1);
    for (const b of playwright.allBrowsers()) {
      if (b !== browser$1)
        await b.close({ reason: "Connection terminated" });
    }
    this._cleanups.push(() => playwrightDispatcher$1.cleanup());
    return playwrightDispatcher$1;
  }
  async _initPreLaunchedAndroidMode(scope) {
    debugLogger.debugLogger.log("server", `[${this._id}] engaged pre-launched (Android) mode`);
    const playwright = this._preLaunched.playwright;
    const androidDevice = this._preLaunched.androidDevice;
    androidDevice.on(android.AndroidDevice.Events.Close, () => {
      this.close({ code: 1001, reason: "Android device disconnected" });
    });
    const playwrightDispatcher$1 = new playwrightDispatcher.PlaywrightDispatcher(scope, playwright, void 0, void 0, androidDevice);
    this._cleanups.push(() => playwrightDispatcher$1.cleanup());
    return playwrightDispatcher$1;
  }
  _initDebugControllerMode() {
    debugLogger.debugLogger.log("server", `[${this._id}] engaged reuse controller mode`);
    const playwright = this._preLaunched.playwright;
    return new debugControllerDispatcher.DebugControllerDispatcher(this._dispatcherConnection, playwright.debugController);
  }
  async _initReuseBrowsersMode(scope) {
    debugLogger.debugLogger.log("server", `[${this._id}] engaged reuse browsers mode for ${this._options.browserName}`);
    const playwright = this._preLaunched.playwright;
    const requestedOptions = launchOptionsHash(this._options.launchOptions);
    let browser$1 = playwright.allBrowsers().find((b) => {
      if (b.options.name !== this._options.browserName)
        return false;
      const existingOptions = launchOptionsHash(b.options.originalLaunchOptions);
      return existingOptions === requestedOptions;
    });
    for (const b of playwright.allBrowsers()) {
      if (b === browser$1)
        continue;
      if (b.options.name === this._options.browserName && b.options.channel === this._options.launchOptions.channel)
        await b.close({ reason: "Connection terminated" });
    }
    if (!browser$1) {
      browser$1 = await playwright[this._options.browserName || "chromium"].launch(instrumentation.serverSideCallMetadata(), {
        ...this._options.launchOptions,
        headless: !!process.env.PW_DEBUG_CONTROLLER_HEADLESS
      });
      browser$1.on(browser.Browser.Events.Disconnected, () => {
        this.close({ code: 1001, reason: "Browser closed" });
      });
    }
    this._cleanups.push(async () => {
      for (const browser2 of playwright.allBrowsers()) {
        for (const context of browser2.contexts()) {
          if (!context.pages().length)
            await context.close({ reason: "Connection terminated" });
          else
            await context.stopPendingOperations("Connection closed");
        }
        if (!browser2.contexts())
          await browser2.close({ reason: "Connection terminated" });
      }
    });
    const playwrightDispatcher$1 = new playwrightDispatcher.PlaywrightDispatcher(scope, playwright, void 0, browser$1);
    return playwrightDispatcher$1;
  }
  async _createOwnedSocksProxy(playwright) {
    if (!this._options.socksProxyPattern)
      return;
    const socksProxy$1 = new socksProxy.SocksProxy();
    socksProxy$1.setPattern(this._options.socksProxyPattern);
    playwright.options.socksProxyPort = await socksProxy$1.listen(0);
    debugLogger.debugLogger.log("server", `[${this._id}] started socks proxy on port ${playwright.options.socksProxyPort}`);
    this._cleanups.push(() => socksProxy$1.close());
    return socksProxy$1;
  }
  async _onDisconnect(error) {
    this._disconnected = true;
    debugLogger.debugLogger.log("server", `[${this._id}] disconnected. error: ${error}`);
    this._root._dispose();
    debugLogger.debugLogger.log("server", `[${this._id}] starting cleanup`);
    for (const cleanup of this._cleanups)
      await cleanup().catch(() => {
      });
    await profiler.stopProfiling(this._profileName);
    this._onClose();
    debugLogger.debugLogger.log("server", `[${this._id}] finished cleanup`);
  }
  logServerMetadata(message, messageString, direction) {
    const serverLogMetadata = {
      wallTime: Date.now(),
      id: message.id,
      guid: message.guid,
      method: message.method,
      payloadSizeInBytes: Buffer.byteLength(messageString, "utf-8")
    };
    debugLogger.debugLogger.log("server:metadata", (direction === "SEND" ? "SEND ► " : "◀ RECV ") + JSON.stringify(serverLogMetadata));
  }
  async close(reason) {
    if (this._disconnected)
      return;
    debugLogger.debugLogger.log("server", `[${this._id}] force closing connection: ${reason?.reason || ""} (${reason?.code || 0})`);
    try {
      this._ws.close(reason?.code, reason?.reason);
    } catch (e) {
    }
  }
}
function launchOptionsHash(options) {
  const copy = { ...options };
  for (const k of Object.keys(copy)) {
    const key = k;
    if (copy[key] === defaultLaunchOptions[key])
      delete copy[key];
  }
  for (const key of optionsThatAllowBrowserReuse)
    delete copy[key];
  return JSON.stringify(copy);
}
function filterLaunchOptions(options, allowFSPaths) {
  return {
    channel: options.channel,
    args: options.args,
    ignoreAllDefaultArgs: options.ignoreAllDefaultArgs,
    ignoreDefaultArgs: options.ignoreDefaultArgs,
    timeout: options.timeout,
    headless: options.headless,
    proxy: options.proxy,
    chromiumSandbox: options.chromiumSandbox,
    firefoxUserPrefs: options.firefoxUserPrefs,
    slowMo: options.slowMo,
    executablePath: debug.isUnderTest() || allowFSPaths ? options.executablePath : void 0,
    downloadsPath: allowFSPaths ? options.downloadsPath : void 0
  };
}
const defaultLaunchOptions = {
  ignoreAllDefaultArgs: false,
  handleSIGINT: false,
  handleSIGTERM: false,
  handleSIGHUP: false,
  headless: true,
  devtools: false
};
const optionsThatAllowBrowserReuse = [
  "headless",
  "tracesDir"
];

exports.PlaywrightConnection = PlaywrightConnection;
