// @cloudflare/playwright-mcp/lib/cjs/src/tools/dialogs.js
'use strict';

Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });

const zod = require('zod');
const tool = require('./tool.js');

const handleDialog = (captureSnapshot) => tool.defineTool({
  capability: "core",
  schema: {
    name: "browser_handle_dialog",
    description: "Handle a dialog",
    inputSchema: zod.z.object({
      accept: zod.z.coerce.boolean().describe("Whether to accept the dialog."),
      promptText: zod.z.string().optional().describe("The text of the prompt in case of a prompt dialog.")
    })
  },
  handle: async (context, params) => {
    const dialogState = context.modalStates().find((state) => state.type === "dialog");
    if (!dialogState)
      throw new Error("No dialog visible");
    if (params.accept)
      await dialogState.dialog.accept(params.promptText);
    else
      await dialogState.dialog.dismiss();
    context.clearModalState(dialogState);
    const code = [
      `// <internal code to handle "${dialogState.dialog.type()}" dialog>`
    ];
    return {
      code,
      captureSnapshot,
      waitForNetwork: false
    };
  },
  clearsModalState: "dialog"
});
const dialogs = (captureSnapshot) => [
  handleDialog(captureSnapshot)
];

exports.default = dialogs;
