// @cloudflare/playwright-mcp/lib/cjs/src/tools/files.js
'use strict';

Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });

const zod = require('zod');
const tool = require('./tool.js');

const uploadFile = (captureSnapshot) => tool.defineTool({
  capability: "files",
  schema: {
    name: "browser_file_upload",
    description: "Upload one or multiple files",
    inputSchema: zod.z.object({
      paths: zod.z.array(zod.z.string()).describe("The absolute paths to the files to upload. Can be a single file or multiple files.")
    })
  },
  handle: async (context, params) => {
    const modalState = context.modalStates().find((state) => state.type === "fileChooser");
    if (!modalState)
      throw new Error("No file chooser visible");
    const code = [
      `// <internal code to chose files ${params.paths.join(", ")}`
    ];
    const action = async () => {
      await modalState.fileChooser.setFiles(params.paths);
      context.clearModalState(modalState);
    };
    return {
      code,
      action,
      captureSnapshot,
      waitForNetwork: true
    };
  },
  clearsModalState: "fileChooser"
});
const files = (captureSnapshot) => [
  uploadFile(captureSnapshot)
];

exports.default = files;
