// @cloudflare/playwright-mcp/lib/cjs/src/tools/navigate.js
'use strict';

Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });

const zod = require('zod');
const tool = require('./tool.js');

const navigate = (captureSnapshot) => tool.defineTool({
  capability: "core",
  schema: {
    name: "browser_navigate",
    description: "Navigate to a URL",
    inputSchema: zod.z.object({
      url: zod.z.string().describe("The URL to navigate to")
    })
  },
  handle: async (context, params) => {
    const tab = await context.ensureTab();
    await tab.navigate(params.url);
    const code = [
      `// Navigate to ${params.url}`,
      `await page.goto('${params.url}');`
    ];
    return {
      code,
      captureSnapshot,
      waitForNetwork: false
    };
  }
});
const goBack = (captureSnapshot) => tool.defineTool({
  capability: "history",
  schema: {
    name: "browser_navigate_back",
    description: "Go back to the previous page",
    inputSchema: zod.z.object({})
  },
  handle: async (context) => {
    const tab = await context.ensureTab();
    await tab.page.goBack();
    const code = [
      `// Navigate back`,
      `await page.goBack();`
    ];
    return {
      code,
      captureSnapshot,
      waitForNetwork: false
    };
  }
});
const goForward = (captureSnapshot) => tool.defineTool({
  capability: "history",
  schema: {
    name: "browser_navigate_forward",
    description: "Go forward to the next page",
    inputSchema: zod.z.object({})
  },
  handle: async (context) => {
    const tab = context.currentTabOrDie();
    await tab.page.goForward();
    const code = [
      `// Navigate forward`,
      `await page.goForward();`
    ];
    return {
      code,
      captureSnapshot,
      waitForNetwork: false
    };
  }
});
const navigate$1 = (captureSnapshot) => [
  navigate(captureSnapshot),
  goBack(captureSnapshot),
  goForward(captureSnapshot)
];

exports.default = navigate$1;
