// @cloudflare/playwright-mcp/lib/cjs/src/tools/pdf.js
'use strict';

Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });

const os = require('node:os');
const path = require('node:path');
const zod = require('zod');
const tool = require('./tool.js');
const utils = require('./utils.js');
const javascript = require('../javascript.js');

const pdf = tool.defineTool({
  capability: "pdf",
  schema: {
    name: "browser_pdf_save",
    description: "Save page as PDF",
    inputSchema: zod.z.object({})
  },
  handle: async (context) => {
    const tab = context.currentTabOrDie();
    const fileName = path.join(os.tmpdir(), utils.sanitizeForFilePath(`page-${(/* @__PURE__ */ new Date()).toISOString()}`)) + ".pdf";
    const code = [
      `// Save page as ${fileName}`,
      `await page.pdf(${javascript.formatObject({ path: fileName })});`
    ];
    return {
      code,
      action: async () => tab.page.pdf({ path: fileName }).then(() => {
      }),
      captureSnapshot: false,
      waitForNetwork: false
    };
  }
});
const pdf$1 = [
  pdf
];

exports.default = pdf$1;
