// @cloudflare/playwright-mcp/lib/cjs/src/tools/snapshot.js
'use strict';

Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });

const path = require('node:path');
const os = require('node:os');
const zod = require('zod');
const utils = require('./utils.js');
const context = require('../context.js');
const javascript = require('../javascript.js');
const tool = require('./tool.js');

const snapshot = tool.defineTool({
  capability: "core",
  schema: {
    name: "browser_snapshot",
    description: "Capture accessibility snapshot of the current page, this is better than screenshot",
    inputSchema: zod.z.object({})
  },
  handle: async (context) => {
    await context.ensureTab();
    return {
      code: [`// <internal code to capture accessibility snapshot>`],
      captureSnapshot: true,
      waitForNetwork: false
    };
  }
});
const elementSchema = zod.z.object({
  element: zod.z.string().describe("Human-readable element description used to obtain permission to interact with the element"),
  ref: zod.z.string().describe("Exact target element reference from the page snapshot")
});
const click = tool.defineTool({
  capability: "core",
  schema: {
    name: "browser_click",
    description: "Perform click on a web page",
    inputSchema: elementSchema
  },
  handle: async (context$1, params) => {
    const tab = context$1.currentTabOrDie();
    const locator = tab.snapshotOrDie().refLocator(params.ref);
    const code = [
      `// Click ${params.element}`,
      `await page.${await context.generateLocator(locator)}.click();`
    ];
    return {
      code,
      action: () => locator.click(),
      captureSnapshot: true,
      waitForNetwork: true
    };
  }
});
const drag = tool.defineTool({
  capability: "core",
  schema: {
    name: "browser_drag",
    description: "Perform drag and drop between two elements",
    inputSchema: zod.z.object({
      startElement: zod.z.string().describe("Human-readable source element description used to obtain the permission to interact with the element"),
      startRef: zod.z.string().describe("Exact source element reference from the page snapshot"),
      endElement: zod.z.string().describe("Human-readable target element description used to obtain the permission to interact with the element"),
      endRef: zod.z.string().describe("Exact target element reference from the page snapshot")
    })
  },
  handle: async (context$1, params) => {
    const snapshot2 = context$1.currentTabOrDie().snapshotOrDie();
    const startLocator = snapshot2.refLocator(params.startRef);
    const endLocator = snapshot2.refLocator(params.endRef);
    const code = [
      `// Drag ${params.startElement} to ${params.endElement}`,
      `await page.${await context.generateLocator(startLocator)}.dragTo(page.${await context.generateLocator(endLocator)});`
    ];
    return {
      code,
      action: () => startLocator.dragTo(endLocator),
      captureSnapshot: true,
      waitForNetwork: true
    };
  }
});
const hover = tool.defineTool({
  capability: "core",
  schema: {
    name: "browser_hover",
    description: "Hover over element on page",
    inputSchema: elementSchema
  },
  handle: async (context$1, params) => {
    const snapshot2 = context$1.currentTabOrDie().snapshotOrDie();
    const locator = snapshot2.refLocator(params.ref);
    const code = [
      `// Hover over ${params.element}`,
      `await page.${await context.generateLocator(locator)}.hover();`
    ];
    return {
      code,
      action: () => locator.hover(),
      captureSnapshot: true,
      waitForNetwork: true
    };
  }
});
const typeSchema = elementSchema.extend({
  text: zod.z.string().describe("Text to type into the element"),
  submit: zod.z.coerce.boolean().optional().describe("Whether to submit entered text (press Enter after)"),
  slowly: zod.z.coerce.boolean().optional().describe("Whether to type one character at a time. Useful for triggering key handlers in the page. By default entire text is filled in at once.")
});
const type = tool.defineTool({
  capability: "core",
  schema: {
    name: "browser_type",
    description: "Type text into editable element",
    inputSchema: typeSchema
  },
  handle: async (context$1, params) => {
    const snapshot2 = context$1.currentTabOrDie().snapshotOrDie();
    const locator = snapshot2.refLocator(params.ref);
    const code = [];
    const steps = [];
    if (params.slowly) {
      code.push(`// Press "${params.text}" sequentially into "${params.element}"`);
      code.push(`await page.${await context.generateLocator(locator)}.pressSequentially(${javascript.quote(params.text)});`);
      steps.push(() => locator.pressSequentially(params.text));
    } else {
      code.push(`// Fill "${params.text}" into "${params.element}"`);
      code.push(`await page.${await context.generateLocator(locator)}.fill(${javascript.quote(params.text)});`);
      steps.push(() => locator.fill(params.text));
    }
    if (params.submit) {
      code.push(`// Submit text`);
      code.push(`await page.${await context.generateLocator(locator)}.press('Enter');`);
      steps.push(() => locator.press("Enter"));
    }
    return {
      code,
      action: () => steps.reduce((acc, step) => acc.then(step), Promise.resolve()),
      captureSnapshot: true,
      waitForNetwork: true
    };
  }
});
const selectOptionSchema = elementSchema.extend({
  values: zod.z.array(zod.z.string()).describe("Array of values to select in the dropdown. This can be a single value or multiple values.")
});
const selectOption = tool.defineTool({
  capability: "core",
  schema: {
    name: "browser_select_option",
    description: "Select an option in a dropdown",
    inputSchema: selectOptionSchema
  },
  handle: async (context$1, params) => {
    const snapshot2 = context$1.currentTabOrDie().snapshotOrDie();
    const locator = snapshot2.refLocator(params.ref);
    const code = [
      `// Select options [${params.values.join(", ")}] in ${params.element}`,
      `await page.${await context.generateLocator(locator)}.selectOption(${javascript.formatObject(params.values)});`
    ];
    return {
      code,
      action: () => locator.selectOption(params.values).then(() => {
      }),
      captureSnapshot: true,
      waitForNetwork: true
    };
  }
});
const screenshotSchema = zod.z.object({
  raw: zod.z.coerce.boolean().optional().describe("Whether to return without compression (in PNG format). Default is false, which returns a JPEG image."),
  element: zod.z.string().optional().transform((v) => v === "null" ? void 0 : v).describe("Human-readable element description used to obtain permission to screenshot the element. If not provided, the screenshot will be taken of viewport. If element is provided, ref must be provided too."),
  ref: zod.z.string().optional().transform((v) => v === "null" ? void 0 : v).describe("Exact target element reference from the page snapshot. If not provided, the screenshot will be taken of viewport. If ref is provided, element must be provided too.")
}).refine((data) => {
  return !!data.element === !!data.ref;
}, {
  message: "Both element and ref must be provided or neither.",
  path: ["ref", "element"]
});
const screenshot = tool.defineTool({
  capability: "core",
  schema: {
    name: "browser_take_screenshot",
    description: `Take a screenshot of the current page. You can't perform actions based on the screenshot, use browser_snapshot for actions.`,
    inputSchema: screenshotSchema
  },
  handle: async (context$1, params) => {
    const tab = context$1.currentTabOrDie();
    const snapshot2 = tab.snapshotOrDie();
    const fileType = params.raw ? "png" : "jpeg";
    const fileName = path.join(os.tmpdir(), utils.sanitizeForFilePath(`page-${(/* @__PURE__ */ new Date()).toISOString()}`)) + `.${fileType}`;
    const options = { type: fileType, quality: fileType === "png" ? void 0 : 50, scale: "css", path: fileName };
    const isElementScreenshot = params.element && params.ref;
    const code = [
      `// Screenshot ${isElementScreenshot ? params.element : "viewport"} and save it as ${fileName}`
    ];
    const locator = params.ref ? snapshot2.refLocator(params.ref) : null;
    if (locator)
      code.push(`await page.${await context.generateLocator(locator)}.screenshot(${javascript.formatObject(options)});`);
    else
      code.push(`await page.screenshot(${javascript.formatObject(options)});`);
    const action = async () => {
      const screenshot2 = locator ? await locator.screenshot(options) : await tab.page.screenshot(options);
      return {
        content: [{
          type: "image",
          data: screenshot2.toString("base64"),
          mimeType: fileType === "png" ? "image/png" : "image/jpeg"
        }]
      };
    };
    return {
      code,
      action,
      captureSnapshot: true,
      waitForNetwork: false
    };
  }
});
const snapshot$1 = [
  snapshot,
  click,
  drag,
  hover,
  type,
  selectOption,
  screenshot
];

exports.default = snapshot$1;
