// @cloudflare/playwright/lib/cjs/playwright-core/src/browserServerImpl.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

require('./server/utils/socksProxy.js');
const playwrightServer = require('./remote/playwrightServer.js');
const helper = require('./server/helper.js');
const instrumentation = require('./server/instrumentation.js');
const playwright = require('./server/playwright.js');
const crypto = require('./server/utils/crypto.js');
const stackTrace = require('./utils/isomorphic/stackTrace.js');
const utilsBundle = require('./utilsBundle.js');

class BrowserServerLauncherImpl {
  constructor(browserName) {
    this._browserName = browserName;
  }
  async launchServer(options = {}) {
    const playwright$1 = playwright.createPlaywright({ sdkLanguage: "javascript", isServer: true });
    const socksProxy = void 0;
    playwright$1.options.socksProxyPort = await socksProxy?.listen(0);
    const metadata = instrumentation.serverSideCallMetadata();
    const browser = await playwright$1[this._browserName].launch(metadata, {
      ...options,
      ignoreDefaultArgs: Array.isArray(options.ignoreDefaultArgs) ? options.ignoreDefaultArgs : void 0,
      ignoreAllDefaultArgs: !!options.ignoreDefaultArgs && !Array.isArray(options.ignoreDefaultArgs),
      env: options.env ? envObjectToArray(options.env) : void 0
    }, toProtocolLogger(options.logger)).catch((e) => {
      const log = helper.helper.formatBrowserLogs(metadata.log);
      stackTrace.rewriteErrorMessage(e, `${e.message} Failed to launch browser.${log}`);
      throw e;
    });
    const path = options.wsPath ? options.wsPath.startsWith("/") ? options.wsPath : `/${options.wsPath}` : `/${crypto.createGuid()}`;
    const server = new playwrightServer.PlaywrightServer({ mode: "launchServer", path, maxConnections: Infinity, preLaunchedBrowser: browser, preLaunchedSocksProxy: socksProxy });
    const wsEndpoint = await server.listen(options.port, options.host);
    const browserServer = new utilsBundle.ws.EventEmitter();
    browserServer.process = () => browser.options.browserProcess.process;
    browserServer.wsEndpoint = () => wsEndpoint;
    browserServer.close = () => browser.options.browserProcess.close();
    browserServer[Symbol.asyncDispose] = browserServer.close;
    browserServer.kill = () => browser.options.browserProcess.kill();
    browserServer._disconnectForTest = () => server.close();
    browserServer._userDataDirForTest = browser._userDataDirForTest;
    browser.options.browserProcess.onclose = (exitCode, signal) => {
      server.close();
      browserServer.emit("close", exitCode, signal);
    };
    return browserServer;
  }
}
function toProtocolLogger(logger) {
  return logger ? (direction, message) => {
    if (logger.isEnabled("protocol", "verbose"))
      logger.log("protocol", "verbose", (direction === "send" ? "SEND ► " : "◀ RECV ") + JSON.stringify(message), [], {});
  } : void 0;
}
function envObjectToArray(env) {
  const result = [];
  for (const name in env) {
    if (!Object.is(env[name], void 0))
      result.push({ name, value: String(env[name]) });
  }
  return result;
}

exports.BrowserServerLauncherImpl = BrowserServerLauncherImpl;
