// @cloudflare/playwright/lib/esm/playwright-core/src/server/dispatchers/localUtilsDispatcher.js
import { Dispatcher } from './dispatcher.js';
import { SdkObject } from '../instrumentation.js';
import { zip, harOpen, harLookup, harClose, harUnzip, tracingStarted, traceDiscarded, addStackToTracingNoReply } from '../localUtils.js';
import { getUserAgent } from '../utils/userAgent.js';
import { deviceDescriptors } from '../deviceDescriptors.js';
import { JsonPipeDispatcher } from './jsonPipeDispatcher.js';
import { ProgressController } from '../progress.js';
import { SocksInterceptor } from '../socksInterceptor.js';
import { WebSocketTransport } from '../../../../cloudflare/webSocketTransport.js';
import { fetchData } from '../utils/network.js';
import { resolveGlobToRegexPattern } from '../../utils/isomorphic/urlMatch.js';

class LocalUtilsDispatcher extends Dispatcher {
  constructor(scope, playwright) {
    const localUtils2 = new SdkObject(playwright, "localUtils", "localUtils");
    const deviceDescriptors$1 = Object.entries(deviceDescriptors).map(([name, descriptor]) => ({ name, descriptor }));
    super(scope, localUtils2, "LocalUtils", {
      deviceDescriptors: deviceDescriptors$1
    });
    this._harBackends = /* @__PURE__ */ new Map();
    this._stackSessions = /* @__PURE__ */ new Map();
    this._type_LocalUtils = true;
  }
  async zip(params) {
    return await zip(this._stackSessions, params);
  }
  async harOpen(params, metadata) {
    return await harOpen(this._harBackends, params);
  }
  async harLookup(params, metadata) {
    return await harLookup(this._harBackends, params);
  }
  async harClose(params, metadata) {
    return await harClose(this._harBackends, params);
  }
  async harUnzip(params, metadata) {
    return await harUnzip(params);
  }
  async tracingStarted(params, metadata) {
    return await tracingStarted(this._stackSessions, params);
  }
  async traceDiscarded(params, metadata) {
    return await traceDiscarded(this._stackSessions, params);
  }
  async addStackToTracingNoReply(params, metadata) {
    return await addStackToTracingNoReply(this._stackSessions, params);
  }
  async connect(params, metadata) {
    const controller = new ProgressController(metadata, this._object);
    controller.setLogName("browser");
    return await controller.run(async (progress) => {
      const wsHeaders = {
        "User-Agent": getUserAgent(),
        "x-playwright-proxy": params.exposeNetwork ?? "",
        ...params.headers
      };
      const wsEndpoint = await urlToWSEndpoint(progress, params.wsEndpoint);
      const transport = await WebSocketTransport.connect(progress, wsEndpoint, { headers: wsHeaders, followRedirects: true, debugLogHeader: "x-playwright-debug-log" });
      const socksInterceptor = new SocksInterceptor(transport, params.exposeNetwork, params.socksProxyRedirectPortForTest);
      const pipe = new JsonPipeDispatcher(this);
      transport.onmessage = (json) => {
        if (socksInterceptor.interceptMessage(json))
          return;
        const cb = () => {
          try {
            pipe.dispatch(json);
          } catch (e) {
            transport.close();
          }
        };
        if (params.slowMo)
          setTimeout(cb, params.slowMo);
        else
          cb();
      };
      pipe.on("message", (message) => {
        transport.send(message);
      });
      transport.onclose = (reason) => {
        socksInterceptor?.cleanup();
        pipe.wasClosed(reason);
      };
      pipe.on("close", () => transport.close());
      return { pipe, headers: transport.headers };
    }, params.timeout || 0);
  }
  async globToRegex(params, metadata) {
    const regex = resolveGlobToRegexPattern(params.baseURL, params.glob, params.webSocketUrl);
    return { regex };
  }
}
async function urlToWSEndpoint(progress, endpointURL) {
  if (endpointURL.startsWith("ws"))
    return endpointURL;
  progress?.log(`<ws preparing> retrieving websocket url from ${endpointURL}`);
  const fetchUrl = new URL(endpointURL);
  if (!fetchUrl.pathname.endsWith("/"))
    fetchUrl.pathname += "/";
  fetchUrl.pathname += "json";
  const json = await fetchData({
    url: fetchUrl.toString(),
    method: "GET",
    timeout: progress?.timeUntilDeadline() ?? 3e4,
    headers: { "User-Agent": getUserAgent() }
  }, async (params, response) => {
    return new Error(`Unexpected status ${response.statusCode} when connecting to ${fetchUrl.toString()}.
This does not look like a Playwright server, try connecting via ws://.`);
  });
  progress?.throwIfAborted();
  const wsUrl = new URL(endpointURL);
  let wsEndpointPath = JSON.parse(json).wsEndpointPath;
  if (wsEndpointPath.startsWith("/"))
    wsEndpointPath = wsEndpointPath.substring(1);
  if (!wsUrl.pathname.endsWith("/"))
    wsUrl.pathname += "/";
  wsUrl.pathname += wsEndpointPath;
  wsUrl.protocol = wsUrl.protocol === "https:" ? "wss:" : "ws:";
  return wsUrl.toString();
}

export { LocalUtilsDispatcher };
